### PR TITLE
Complete route blinding support

### PR DESCRIPTION
**Pull request by valentinewallace**
We'll now support forwarding blinded payments as not-the-intro-node. Unblocks advertising the feature bit, though we don't advertise it yet in this PR. 

Helps address #1970. There are still a few small follow-ups planned after this one.

### CODE REVIEW

File: lightning/src/ln/channelmanager.rs [hunk 1]
```diff
@@ -9481,6 +9488,7 @@ impl_writeable_tlv_based!(PhantomRouteHints, {
 
 impl_writeable_tlv_based!(BlindedForward, {
 	(0, inbound_blinding_point, required),
+	(2, failure, (default_value, BlindedFailure::FromIntroductionNode)),
```
[the remark this answers is not in the document]
Comment (valentinewallace) [reply lightning/src/ln/channelmanager.rs]: Good catch, prefer the latter. Fixed.

File: lightning/src/ln/channelmanager.rs [hunk 2]
```diff
@@ -201,7 +201,9 @@ pub struct BlindedForward {
 	/// onion payload if we're the introduction node. Useful for calculating the next hop's
 	/// [`msgs::UpdateAddHTLC::blinding_point`].
 	pub inbound_blinding_point: PublicKey,
-	// Another field will be added here when we support forwarding as a non-intro node.
+	/// If needed, this determines how this HTLC should be failed backwards, based on whether we are
+	/// the introduction node.
+	pub failure: BlindedFailure,
```
[the remark this answers is not in the document]
Comment (valentinewallace) [reply lightning/src/ln/channelmanager.rs:206]: Hm, it should only be used in a failure context. Could you point me to where you're seeing that?

File: lightning/src/ln/blinded_payment_tests.rs [hunk 3]
```diff
@@ -1050,3 +1050,100 @@ fn high_prop_fees() {
...
+
+#[cfg(feature = "std")]
+#[test]
+fn prop_fees_rng() {
+	do_prop_fees_rng(true);
+	do_prop_fees_rng(false);
+}
+
+#[cfg(feature = "std")]
+fn do_prop_fees_rng(send_min: bool) {
```
[the remark this answers is not in the document]
Comment (valentinewallace) [reply lightning/src/ln/blinded_payment_tests.rs:1062]: This was written a long time ago to have actual testing for #2514. The random values helped catch some edge cases we weren't handling.

**Comment by valentinewallace** [comment]
Because we may have a shorter than usual release cycle for 0.0.120, I moved the tests and the related test util refactors into #2823 in case it gives us a better chance of landing this in time.

File: lightning/src/ln/functional_test_utils.rs [hunk 4]
```diff
@@ -2576,8 +2576,11 @@ pub fn pass_claimed_payment_along_route<'a, 'b, 'c>(origin_node: &Node<'a, 'b, '
 		per_path_msgs.push(msgs_from_ev!(&events[0]));
 	} else {
 		for expected_path in expected_paths.iter() {
-			// For MPP payments, we always want the message to the first node in the path.
-			let ev = remove_first_msg_event_to_node(&expected_path[0].node.get_our_node_id(), &mut events);
+			// For MPP payments, we always want the message to the first node in the path back to the
+			// sender.
+			let ev = remove_first_msg_event_to_node(
+				&expected_path[expected_path.len() - 2].node.get_our_node_id(), &mut events
```
[the remark this answers is not in the document]
Comment (valentinewallace) [reply lightning/src/ln/functional_test_utils.rs]: I updated the comment in #2823.

File: lightning/src/ln/channelmanager.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkczyz) [reply lightning/src/ln/channelmanager.rs:206]: By "used" I meant more that it is used in this struct even though described "if needed". I suppose the struct name is fine given how it is actually used, but the variable name `failure` makes it seem like something has already failed when in fact in this context it might not ever. I'm fine if there's not a better name.

**Comment by valentinewallace** [comment]
Rebased due to conflicts.

File: lightning/src/ln/onion_payment.rs [hunk 5]
```diff
@@ -105,7 +105,12 @@ pub(super) fn create_fwd_pending_htlc_info(
 		routing: PendingHTLCRouting::Forward {
 			onion_packet: outgoing_packet,
 			short_channel_id,
-			blinded: inbound_blinding_point.map(|bp| BlindedForward { inbound_blinding_point: bp }),
+			blinded: intro_node_blinding_point.or(msg.blinding_point)
+				.map(|bp| BlindedForward {
+					inbound_blinding_point: bp,
+					failure: intro_node_blinding_point
+						.map_or(BlindedFailure::FromBlindedNode, |_| BlindedFailure::FromIntroductionNode),
```
[the remark this answers is not in the document]
Comment (valentinewallace) [reply lightning/src/ln/onion_payment.rs:112]: `map_or` syntax is a bit weird but `BlindedFailure::FromBlindedNode` is the default value that gets returned if `intro_node_blinding_point` is `None`. 

> Unrelatedly, FromIntroductionNode should probably read WereIntroductionNode, right?

Hmm... that reads to me as "were introduction node." Do you think the current name is unclear?

File: lightning/src/ln/channelmanager.rs [hunk 6]
```diff
@@ -9495,6 +9503,7 @@ impl_writeable_tlv_based!(PhantomRouteHints, {
 
 impl_writeable_tlv_based!(BlindedForward, {
 	(0, inbound_blinding_point, required),
+	(1, failure, (default_value, BlindedFailure::FromIntroductionNode)),
```
[the remark this answers is not in the document]
Comment (valentinewallace) [reply lightning/src/ln/channelmanager.rs:9506]: Ah, no we previously only supported intro node forwarding.

File: lightning/src/ln/onion_payment.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jkczyz) [reply lightning/src/ln/onion_payment.rs:112]: > > map_or syntax is a bit weird but BlindedFailure::FromBlindedNode is the default value that gets returned if intro_node_blinding_point is None.
> 
> Ugh.
> 

Agreed regarding `map_or` parameters. Might be clearer as:

```rust
intro_node_blinding_point.map(|_| BlindedFailure::FromIntroductionNode).or(BlindedFailure::FromBlindedNode)
```

**Review by jkczyz — COMMENTED** [review]
LGTM. Feel free to squash.

File: lightning/src/ln/onion_payment.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (valentinewallace) [reply lightning/src/ln/onion_payment.rs]: Had to use `unwrap_or` but did this.

**Review by jkczyz — APPROVED** [review]
